crsqlite_extension/src/storage.c

```diff
@@ -0,0 +1,106 @@
+#include <stdio.h>
+#include <string.h>
+#include <stdarg.h>
+
+#include <sqlite3ext.h>
+
+#ifndef sqlite3_api
+SQLITE_EXTENSION_INIT3
+#endif
+
+#define COLUMN_NAME 1
+#define COLUMN_TYPE 2
+#define COLUMN_NOT_NULL 3
+#define COLUMN_DFLT_VAL 4
+#define COLUMN_PK 5
+
+ /*
+** Construct one or more SQL statements from the format string given
+** and then evaluate those statements. The success code is written
+** into *pRc.
+**
+** If *pRc is initially non-zero then this routine is a no-op.
+*/
+static void storageDbExec(
+  int *pRc,              /* Success code */
+  sqlite3 *db,           /* Database in which to run SQL */
+  const char *zFormat,   /* Format string for SQL */
+  ...                    /* Arguments to the format string */
+){
+  va_list ap;
+  char *zSql;
+  if( *pRc ) return;
+  va_start(ap, zFormat);
+  zSql = sqlite3_vmprintf(zFormat, ap);
+  //printf("Executing SQL statement: %s\n", zSql);
+  va_end(ap);
+  if( zSql==0 ){
+    *pRc = SQLITE_NOMEM;
+    
+  }else{
+    *pRc = sqlite3_exec(db, zSql, 0, 0, 0);
+    sqlite3_free(zSql);
+  }
+}
+
+
+/* Initites the real database table(s) that backs the virtual table*/
+int init_storage(
+  sqlite3 *db,
+  const char* ttbl,
+  const char* createTableArgs,
+  char **pzErr
+)
+{
+  
+  //Create the table with "normal" schema
+  int rc = SQLITE_OK;
+  storageDbExec(&rc, db, "CREATE TABLE crsqlite_%s(%s);", ttbl, createTableArgs);
+  if (rc != SQLITE_OK){
+    fprintf(stderr,"Error creating table: %s\n", ttbl);
+    return rc;
+  }
+
+  char *crsqliteTableName = sqlite3_mprintf("crsqlite_%s", ttbl);
+  if (crsqliteTableName == NULL) return SQLITE_NOMEM;
+  char *zSql = sqlite3_mprintf("PRAGMA table_info('%s');", crsqliteTableName);
+  if (zSql == NULL) return SQLITE_NOMEM;
+  sqlite3_stmt *pStmt = 0;
+  rc = sqlite3_prepare(db, zSql, -1, &pStmt, 0);
+  sqlite3_free(zSql);
+
+  //Excute table_info pragma on table:
```

Comment:
what's your thinking around supporting alter table statements?

I imagine we'll need to understand how to parse `createTableArgs` directly rather than relying on creating the table then using the pragma. Any other thoughts?

Fortunately, column defs look pretty simple (https://www.sqlite.org/syntax/column-def.html) until you hit the column constraint (https://www.sqlite.org/syntax/column-constraint.html) but I think we don't have to fully parse the column constraint piece.

Reply:
Good idea, as discussed, this was a bit of a lazy approach to get used to be able to create tables for now.

Reply:
May I suggest to take a look at my repo? https://github.com/marcobambini/sqlite-createtable-parser

Reply:
Thanks for the tip @marcobambini , I'll have a look :-)